main_test.go

```diff
@@ -1,62 +1,198 @@
-			signed = signCookie("user,group", secret) + "garbage"
-		})
+	var nonce = ""
+	if options.Nonce != "" {
+		nonce = options.Nonce
+	} else {
+		nonce = addNonce("http://some.url/")
+	}
 
-		It("fails", func() {
-			Expect(parseError).To(HaveOccurred())
-		})
-	})
+	ssoEncoded := base64.StdEncoding.EncodeToString([]byte(sso + nonce))
+
+	queryString := u.Query()
+	queryString.Set("sso", ssoEncoded)
+	queryString.Set("sig", computeHMAC(ssoEncoded, options.Secret))
+
+	u.RawQuery = queryString.Encode()
+
+	return u.String()
+}
+
+type SSOOverrideFunction func(*SSOOptions)
+type ConfigOverrideFunction func(*Config)
 
-	Context("when verifying with a valid payload and secret", func() {
-		BeforeEach(func() {
-			secret = "mysecret"
-			signed = signCookie("user,group", secret)
-		})
-
-		It("returns a user and group", func() {
-			Expect(parseError).To(Succeed())
-			Expect(parsedUsername).To(Equal("user"))
-			Expect(parsedGroup).To(Equal("group"))
-		})
+func GetTestResult(configOverride ConfigOverrideFunction, ssoOverride SSOOverrideFunction) *http.Response {
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintln(w, "")
 	})
-})
+
+	config := GetTestConfig()
+
+	configOverride(&config)
+	setConfig(&config)
+
+	proxy := authProxyHandler(handler)
+	ts := httptest.NewServer(proxy)
+	defer ts.Close()
+
+	options := GetSSOOptions(ts.URL, config.SSOSecret)
+	ssoOverride(&options)
+
+	res, _ := http.Get(GetTestSSOURL(options))
+	return res
+}
+
+func TestBadSecret(t *testing.T) {
+
+	res := GetTestResult(
+		func(config *Config) {
+			config.AllowAll = true
+		},
+		func(options *SSOOptions) {
+			options.Secret = "BAD SECRET"
+		},
+	)
+
+	assert.Equal(t, 400, res.StatusCode)
+}
+
+func TestWrongGroup(t *testing.T) {
+
+	res := GetTestResult(
+		func(config *Config) {
+			config.AllowGroups = "group_a,group_b"
+		},
+		func(options *SSOOptions) {
+			options.Groups = "group_c,group_d"
+		},
+	)
+
+	assert.Equal(t, 403, res.StatusCode)
+}
+
+func TestRightGroup(t *testing.T) {
+
+	res := GetTestResult(
+		func(config *Config) {
+			config.AllowGroups = "group_a,group_b"
+		},
+		func(options *SSOOptions) {
+			options.Groups = "group_c,group_a"
+		},
+	)
+
+	assert.Equal(t, 200, res.StatusCode)
+}
+
+func TestNoRandomAnonAllowed(t *testing.T) {
+
+	res := GetTestResult(
+		func(config *Config) {
+			config.AllowGroups = ""
+			config.AllowAll = false
+		},
+		func(options *SSOOptions) {
+			options.Admin = false
+		},
+	)
+
+	assert.Equal(t, 403, res.StatusCode)
+}
+
+func TestAllowedAnon(t *testing.T) {
+	res := GetTestResult(
+		func(config *Config) {
+			config.AllowGroups = ""
+			config.AllowAll = true
+		},
+		func(options *SSOOptions) {
+			options.Admin = false
+		},
+	)
+
+	assert.Equal(t, 200, res.StatusCode)
+}
+
+func TestInvalidSecretFails(t *testing.T) {
+	signed := signCookie("user,group", "secretfoo")
+	_, _, parseError := parseCookie(signed, "secretbar")
+
+	assert.NotNil(t, parseError)
```

Comment:
I think [`assert.Error`](https://pkg.go.dev/github.com/stretchr/testify@v1.6.1/assert#Error) would be more obvious.

Reply:
I think you had intended to change this?  (`NotNil` &rarr; `Error`.  That change was made in one place below, but not here.)